File: 02_Data_Structures/trees/heaps/min_heap/implementations/operations.cpp
/*
----------------------------------------
Min Heap Operations Implementation (C++)
File: /02_Data_Structures/trees/heaps/min_heap/implementations/operations.cpp
----------------------------------------

A Min Heap is a complete binary tree where:
- Every parent node is <= its children.
- Smallest element is always at root.

This file implements:
- insert(x)
- extractMin()
- heapifyDown()
- heapifyUp()

Time Complexity:
- insert: O(log n)
- extractMin: O(log n)
- heapify/buildHeap: O(n)
Space Complexity: O(n)

Use-case:
- Priority queues
- Dijkstra, Prim
- Scheduling
*/

#include <bits/stdc++.h>
using namespace std;

class MinHeap {
private:
    vector<int> h;

    // Get parent index
    int parent(int i) { return (i - 1) / 2; }
    int left(int i) { return 2 * i + 1; }
    int right(int i) { return 2 * i + 2; }

    // Maintains heap property upward after insertion
    void heapifyUp(int i) {
        while (i > 0 && h[parent(i)] > h[i]) {
            swap(h[i], h[parent(i)]);
            i = parent(i);
        }
    }

    // Maintains heap property downward after extractMin
    void heapifyDown(int i) {
        int n = h.size();
        while (true) {
            int l = left(i), r = right(i);
            int smallest = i;

            if (l < n && h[l] < h[smallest]) smallest = l;
            if (r < n && h[r] < h[smallest]) smallest = r;

            if (smallest == i) break;

            swap(h[i], h[smallest]);
            i = smallest;
        }
    }

public:
    // Insert element
    void insert(int x) {
        h.push_back(x);
        heapifyUp(h.size() - 1);
    }

    // Returns minimum element
    int getMin() {
        if (h.empty()) throw runtime_error("Heap is empty");
        return h[0];
    }

    // Extract minimum and fix heap
    int extractMin() {
        if (h.empty()) throw runtime_error("Heap is empty");

        int mn = h[0];
        h[0] = h.back();
        h.pop_back();
        if (!h.empty()) heapifyDown(0);

        return mn;
    }

    // Print heap as array (level order)
    void printHeap() {
        for (int x : h) cout << x << " ";
        cout << "\n";
    }
};

int main() {
    MinHeap mh;
    mh.insert(10);
    mh.insert(5);
    mh.insert(20);
    mh.insert(2);
    mh.insert(15);

    cout << "Heap: ";
    mh.printHeap();

    cout << "Extract Min: " << mh.extractMin() << "\n";

    cout << "Heap after extraction: ";
    mh.printHeap();

    return 0;
}

File: 02_Data_Structures/trees/heaps/min_heap/implementations/operations.py
# Min Heap Operations Implementation (Python)
# File: /02_Data_Structures/trees/heaps/min_heap/implementations/operations.py
# ------------------------------------------------------------
#
# A Min Heap is a complete binary tree represented as an array where:
# - h[i] <= h[left(i)]
# - h[i] <= h[right(i)]
# - Root (index 0) always contains the minimum element.
#
# Supported operations:
# - insert(x)
# - extract_min()
# - get_min()
# - heapify_up()
# - heapify_down()
#
# Time Complexity:
# insert: O(log n)
# extract_min: O(log n)
# heapify/buildHeap: O(n)
# Space Complexity: O(n)
#
# Use cases:
# - Priority queues
# - Dijkstra's algorithm
# - Prim's MST
# - Anytime you need quick access to the minimum element
# ------------------------------------------------------------

class MinHeap:
    def __init__(self):
        self.h = []  # underlying array

    # -----------------------
    # Utility index functions
    # -----------------------
    def parent(self, i):
        return (i - 1) // 2

    def left(self, i):
        return 2 * i + 1

    def right(self, i):
        return 2 * i + 2

    # -----------------------
    # Heapify Up (after insert)
    # -----------------------
    def heapify_up(self, i):
        while i > 0 and self.h[self.parent(i)] > self.h[i]:
            self.h[self.parent(i)], self.h[i] = self.h[i], self.h[self.parent(i)]
            i = self.parent(i)

    # -----------------------
    # Heapify Down (after extract_min)
    # -----------------------
    def heapify_down(self, i):
        n = len(self.h)
        while True:
            l = self.left(i)
            r = self.right(i)
            smallest = i

            if l < n and self.h[l] < self.h[smallest]:
                smallest = l
            if r < n and self.h[r] < self.h[smallest]:
                smallest = r

            if smallest == i:
                break

            self.h[i], self.h[smallest] = self.h[smallest], self.h[i]
            i = smallest

    # -----------------------
    # Insert element into heap
    # -----------------------
    def insert(self, x):
        self.h.append(x)
        self.heapify_up(len(self.h) - 1)

    # -----------------------
    # Get min element
    # -----------------------
    def get_min(self):
        if not self.h:
            raise IndexError("Heap is empty")
        return self.h[0]

    # -----------------------
    # Extract min + rearrange heap
    # -----------------------
    def extract_min(self):
        if not self.h:
            raise IndexError("Heap is empty")

        root = self.h[0]
        self.h[0] = self.h[-1]
        self.h.pop()

        if self.h:
            self.heapify_down(0)

        return root

    # -----------------------
    # Print heap (level-order)
    # -----------------------
    def print_heap(self):
        print("Heap:", self.h)


# ------------------------------------------------------------
# Example usage
# ------------------------------------------------------------
if __name__ == "__main__":
    mh = MinHeap()
    mh.insert(10)
    mh.insert(4)
    mh.insert(15)
    mh.insert(1)
    mh.insert(20)

    mh.print_heap()

    print("Minimum:", mh.get_min())
    print("Extracted:", mh.extract_min())

    mh.print_heap()

